airflow/contrib/operators/gcs_copy_operator.py

```diff
@@ -0,0 +1,115 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from airflow.contrib.hooks.gcs_hook import GoogleCloudStorageHook
+from airflow.models import BaseOperator
+from airflow.utils.decorators import apply_defaults
+
+
+class GoogleCloudStorageCopyOperator(BaseOperator):
+    """
+    Copies objects (optionally from a directory) filtered by 'delimiter' (file extension for e.g .json) from a bucket
+    to another bucket in a different directory, if required.
+
+    :param source_bucket: The source Google cloud storage bucket where the object is.
+    :type source_bucket: string
+    :param source_object: The source name of the object to copy in the Google cloud
+        storage bucket.
+    :type source_object: string
+    :param source_files_delimiter: The delimiter by which you want to filter the files to copy.
+        For e.g to copy the CSV files from in a directory in GCS you would use source_files_delimiter='.csv'.
+    :type source_files_delimiter: string
+    :param destination_bucket: The destination Google cloud storage bucket where the object should be.
+    :type destination_bucket: string
+    :param destination_directory: The destination name of the directory in the destination Google cloud
+        storage bucket.
+    :type destination_directory: string
+    :param google_cloud_storage_conn_id: The connection ID to use when
+        connecting to Google cloud storage.
+    :type google_cloud_storage_conn_id: string
+    :param delegate_to: The account to impersonate, if any.
+        For this to work, the service account making the request must have domain-wide delegation enabled.
+    :type delegate_to: string
+
+    Example: The following Operator would move all the CSV files from `sales/sales-2017` folder in `data` bucket to
+    `sales` folder in `archive` bucket.
+
+    move_file = GoogleCloudStorageCopyOperator(
+        task_id='move_file',
+        source_bucket='data',
+        source_object='sales/sales-2017/',
+        source_files_delimiter='.csv'
+        destination_bucket='archive',
+        destination_directory='sales',
+        google_cloud_storage_conn_id='airflow-service-account'
+    )
+    """
+    template_fields = ('source_bucket', 'source_object', 'source_files_delimiter',
+                       'destination_bucket', 'destination_directory')
+    ui_color = '#f0eee4'
+
+    @apply_defaults
+    def __init__(self,
+                 source_bucket,
+                 source_object,
```

Comment:
better call it prefix than source_object according to your example

Reply:
Can do that. But the notion was that even if you want to copy a file "abc.txt" you would pass source_object="abc.txt". If instead of `source_object` we call it `prefix` than the user might not find `prefix` keyword intuitive enough.

But yes if `prefix` seems logical than we can change it. Would like to have the opinion of others as well.